init-tools.sh

```diff
@@ -99,14 +99,30 @@ if [ ! -e $__DOTNET_PATH ]; then
     echo "Installing dotnet cli..."
     __DOTNET_LOCATION="https://dotnetcli.azureedge.net/dotnet/Sdk/${__DOTNET_TOOLS_VERSION}/${__DOTNET_PKG}.tar.gz"
     # curl has HTTPS CA trust-issues less often than wget, so lets try that first.
-    echo "Installing '${__DOTNET_LOCATION}' to '$__DOTNET_PATH/dotnet.tar'" >> $__init_tools_log
-    if command -v curl > /dev/null; then
-        curl --retry 10 -sSL --create-dirs -o $__DOTNET_PATH/dotnet.tar ${__DOTNET_LOCATION}
-    else
-        wget -q -O $__DOTNET_PATH/dotnet.tar ${__DOTNET_LOCATION}
-    fi
-    cd $__DOTNET_PATH
-    tar -xf $__DOTNET_PATH/dotnet.tar
+    __retryCount=5
+    __retryWaitFactor=6 
+    __count=0
+    until [ $__count -ge $__retryCount ]
+    do
+        echo "Installing '${__DOTNET_LOCATION}' to '$__DOTNET_PATH/dotnet.tar'" >> $__init_tools_log
+        if command -v curl > /dev/null; then
+            curl --retry 10 -sSL --create-dirs -o $__DOTNET_PATH/dotnet.tar ${__DOTNET_LOCATION}
+        else
+            wget -q -O $__DOTNET_PATH/dotnet.tar ${__DOTNET_LOCATION}
+        fi
+        cd $__DOTNET_PATH
+        if tar -xf $__DOTNET_PATH/dotnet.tar; then break; fi
+
+        # wait and retry if tar command fails
+        __waitTime=$(( __retryWaitFactor ** __count ))
+        __count=$(( __count + 1 ))
+        if [ $__count -ne $__retryCount ]; then
+            echo "Install failed.  Waiting for ${__waitTime} seconds and trying again" >> $__init_tools_log
+            sleep $__waitTime
+        else
+            echo "Max retries (${retryCount}) reached.  Giving up." >> $__init_tools_log
+        fi
+    done
```

Comment:
I'm not sure what the correct way would be to make the scripts share the same code, but maybe this could use the [retry function we've been using in `init-docker.sh`](https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/scripts/docker/init-docker.sh#L25-L42)? Copying over the function code would at least make it easy to centralize if there is ever a clear way, and we can retry other things in init-tools the same way.

Reply:
I copied the function code and used it for the retry.  Thanks!